========
Overview
========

.. image:: https://readthedocs.org/projects/omicscope/badge/?style=flat
    :target: https://omicscope.readthedocs.io/
    :alt: Documentation Status

.. image:: https://img.shields.io/pypi/v/omicscope.svg
    :alt: PyPI Package latest release
    :target: https://pypi.org/project/omicscope

.. image:: https://img.shields.io/pypi/wheel/omicscope.svg
    :alt: PyPI Wheel
    :target: https://pypi.org/project/omicscope

.. image:: https://img.shields.io/pypi/pyversions/omicscope.svg
    :alt: Supported versions
    :target: https://pypi.org/project/omicscope

.. image:: https://img.shields.io/pypi/implementation/omicscope.svg
    :alt: Supported implementations
    :target: https://pypi.org/project/omicscope

.. image:: https://img.shields.io/github/commits-since/guireiso/omicscope/v1.3.9.svg
    :alt: Commits since latest release
    :target: https://github.com/guireiso/omicscope/compare/v1.3.9...main


OmicScope: OmicScope: from quantitative proteomics to systems biology.
----------------------------------------------------------------------

*OmicScope* is a comprehensive workflow designed to analyze and provide insights on quantitative proteomics data. To date, *OmicScope* works with data generated from `Progenesis QI for Proteomics <https://www.nonlinear.com/progenesis/qi-for-proteomics/>`__, `MaxQuant <https://www.maxquant.org/>`__, and `PatternLab V <http://www.patternlabforproteomics.org/>`__. Additionally, a fourth generic input can be used, enabling users to run *OmicScope* with data from different platforms, including transcriptomics and metabolomics.

*OmicScope* can perform differential expression analysis in both static and longitudinal experimental designs. For static experiments, proteins that are differentially regulated are determined by t-tests (for 2 group comparison) or One-way ANOVA (for >2 group comparison); while for longitudinal analysis, *OmicScope* performs the pipeline suggested by Storey, 2005.

Once the differential expression data is obtained, the user can perform Over-Representation Analysis (ORA) or Gene-Set Enrichment Analysis (GSEA), which are implemented according to `GSEApy <https://gseapy.readthedocs.io/>`__ in the EnrichmentScope module.

Both differential expression and enrichment analyses have a comprehensive toolset of figures generated by *OmicScope*, including dotplots, networks, heatmaps, etc.

Finally, for each experiment performed by *OmicScope*, it is possible to export an omics file (.omics extension) for further use in the **Nebula workflow**. Nebula is a specialized module that enables analysis of multiple studies or comparisons, providing several types of analysis to compare those groups and find similarities among them.


.. image:: ./Workflow.png
   :width: 600
   :align: center

Installation
------------

::

    pip install omicscope

You can also install the in-development version with::

    pip install https://github.com/guireiso/omicscope/archive/main.zip


Documentation
-------------


https://omicscope.readthedocs.io/

