========
Overview
========



.. image:: https://readthedocs.org/projects/omicscope/badge/?style=flat
    :target: https://omicscope.readthedocs.io/
    :alt: Documentation Status

.. image:: https://github.com/guireiso/omicscope/actions/workflows/github-actions.yml/badge.svg
    :alt: GitHub Actions Build Status
    :target: https://github.com/guireiso/omicscope/actions

.. image:: https://img.shields.io/pypi/v/omicscope.svg
    :alt: PyPI Package latest release
    :target: https://pypi.org/project/omicscope

.. image:: https://img.shields.io/pypi/wheel/omicscope.svg
    :alt: PyPI Wheel
    :target: https://pypi.org/project/omicscope

.. image:: https://img.shields.io/pypi/pyversions/omicscope.svg
    :alt: Supported versions
    :target: https://pypi.org/project/omicscope

.. image:: https://img.shields.io/pypi/implementation/omicscope.svg
    :alt: Supported implementations
    :target: https://pypi.org/project/omicscope

.. image:: https://img.shields.io/github/commits-since/guireiso/omicscope/v1.0.1.svg
    :alt: Commits since latest release
    :target: https://github.com/guireiso/omicscope/compare/v1.0.1...main


OmicScope: from DIA proteomics to Systems Biology visualisation
---------------------------------------------------------------

*OmicScope* is a comprehensive workflow designed to analyse and provide insights regarding quantitative proteomics data. Todate, *OmicScope* works with data generated from `Progenesis QI for Proteomics <https://www.nonlinear.com/progenesis/qi-for-proteomics/>`__, `MaxQuant <https://www.maxquant.org/>`__, and `PatternLab V <http://www.patternlabforproteomics.org/>`__. Additionally, a fourth generical input can be used for further analysis, allowing users to run software coming from different plataforms (including transcriptomics and metabolomics).

*OmicScope* can perform differential expression analysis in both static and longitudinal experimental designs. For static experiments, proteins differentially regulated are determined by t-tests (2 group comparison) or One-way ANOVA  (>2 group comparison); while for longitudinal analysis, OmicScope perform the pipeline suggested by `Storey, 2005 <https://www.pnas.org/doi/10.1073/pnas.0504609102>`__. 

Once obtained differential expression data, user can perform Over-Representation Analysis (ORA) or Gene-Set Enrichment Analysis (GSEA), which were implementated  according to `GSEApy <https://gseapy.readthedocs.io/>`__.

Both diferential expression and enrichment analyses have a compreehensive toolset of figures generated by *OmicScope*, including dotplots, networks, heatmaps, etc.

Finally, for each experiment performed by *OmicScope*, it is possible to download the an omics file (.omics extension) for further use in **Nebula workflow**. Nebula is an specialized module that unables analysis of multiple studies or comparions, providing several types of analysis to compare those groups and find similarities among them.


Installation
------------

::

    pip install omicscope

You can also install the in-development version with::

    pip install https://github.com/guireiso/omicscope/archive/main.zip


Documentation
-------------


https://omicscope.readthedocs.io/

